compiler/codegen/mli.cpp

```diff
@@ -0,0 +1,912 @@
+  // Always declare a variable to catch socket errors.
+  gen += this->genNewDecl("int", "skt_err");
+
+  // If unpacking, declare a temporary for the return value.
+  if (not out) {
+    gen += this->genNewDecl(t, "result");
+  }
+
+  // If allocating, declare a temporary for amount.
+  if (this->typeRequiresAllocation(t)) {
+    gen += this->genNewDecl("uint64_t", "bytes");
+  }
+
+  // Insert a debug message if appropriate.
+  if (this->injectDebugPrintlines) {
+    std::string msg;
+
+    msg += out ? "Pushing type: " : "Pulling type: ";
+    msg += this->genTypeName(t);
+
+    gen += this->genDebugPrintCall(msg.c_str(), "[RPC]");
+  }
+
+  //
+  // Handle translation of different type classes here. Note that right now
+  // the only things we can translate are primitive scalars and cstrings.
+  //
+  if (isPrimitiveScalar(t)) {
+    gen += this->genMarshalBodyPrimitiveScalar(t, out);
+  } else if (t == dtStringC) {
+    gen += this->genMarshalBodyStringC(t, out);
+  } else {
+    USR_FATAL("MLI does not support code generation for type", t);
+  }
+
+  // If we are unpacking, return our temporary.
+  if (not out) { gen += "return result;\n"; }
+  
+  gen += scope_end;
+  gen += "\n";
+
+  return gen;
+}
+
+std::string MLIContext::genMarshalPushRoutine(Type* t) {
+  return this->genMarshalRoutine(t, true);
+}
+
+std::string MLIContext::genMarshalPullRoutine(Type* t) {
+  return this->genMarshalRoutine(t, false);
+}
+
+void MLIContext::emitServerDispatchRoutine(void) {
+  std::string gen;
+
+  gen += this->genServerDispatchSwitch(this->exps);
+  gen += "\n";
+  
+  this->setOutputAndWrite(&this->fiServerBundle, gen);
+
+  return;
+}
+
+void MLIContext::setOutput(fileinfo* fi) {
+  this->info->cfile = fi->fptr;
+  return;
+}
+
+void MLIContext::setOutputAndWrite(fileinfo* fi, const std::string& gen) {
+  this->setOutput(fi);
+  this->write(gen);
+}
+
+void MLIContext::write(const std::string& gen) {
+  fprintf(this->info->cfile, gen.c_str());
+  return;
+}
+
+//
+// We can (as I understand it) cound on Type* being unique across the entire
+// symbol table (IE, you'll never encounter two different Type objects that
+// both end up describing the same concrete type).
+//
+int64_t MLIContext::assignUniqueTypeID(Type* t) {
+  // Prepare a new ID based on the map size (will never overflow).
+  int64_t result = (int64_t) this->typeMap.size();
+
+  if (this->typeMap.find(t) != this->typeMap.end()) {
+    result = this->typeMap[t];
+  } else {
+    this->typeMap[t] = result;
+  }
+
+  return result;
+}
+
+void MLIContext::emitClientWrapper(FnSymbol* fn) {
+  std::string gen;
+
+  this->setOutput(&this->fiClientBundle);
+  fn->codegenHeaderC();
+
+  gen += scope_begin;
+  gen += this->genClientsideRPC(fn);
+  gen += scope_end;
+  gen += "\n";
+
+  this->write(gen);
+  
+  return;
+}
+
+void MLIContext::emitServerWrapper(FnSymbol* fn) {
+  std::string gen;
+
+  // Big long, silly block of manual code generation.
+  gen += this->genComment(toString(fn));
```

Comment:
That seems like it could quickly double the size of the generate code, do we need to do this?

Reply:
I don't follow, you don't want debug messages? This just prints out the function's name so it's easier to identify (the wrapper is mangled).

Reply:
Oh, I misunderstood.  I thought toString(fn) printed out more than just the name (like, it printed out the whole function into the comments or something), that's okay then.